### PR TITLE
[SPARK-45574][SQL] Add :: syntax as a shorthand for casting

### PR DESCRIPTION
**Pull request by mitkedb**
### What changes were proposed in this pull request?

Adds the `::` syntax as syntactic sugar for casting columns. This is a pretty common syntax across many industry databases.

### Does this PR introduce _any_ user-facing change?

Yes, new casting syntax.

### How was this patch tested?

Unit tests.
SQL tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]

**Review by peter-toth — APPROVED** [review]

File: sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 1]
```diff
@@ -957,6 +957,7 @@ primaryExpression
     | CASE whenClause+ (ELSE elseExpression=expression)? END                                   #searchedCase
     | CASE value=expression whenClause+ (ELSE elseExpression=expression)? END                  #simpleCase
     | name=(CAST | TRY_CAST) LEFT_PAREN expression AS dataType RIGHT_PAREN                     #cast
+    | primaryExpression DOUBLE_COLON dataType                                                  #castByColon
```
Comment (MaxGekk) [inline sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:960]: Could you clarify, please, don't you allow only `primaryExpression` but not `expression` like at the line above.

Comment (cloud-fan) [reply sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:960]: +1. CAST is also using `expression`

File: sql/core/src/test/resources/sql-tests/inputs/cast_op.sql [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+SELECT NULL :: int;
+SELECT NULL :: long;
+
+-- cast invalid decimal string to numeric types
+SELECT '123.a' :: int;
+SELECT '123.a' :: long;
+SELECT '123.a' :: float;
+SELECT '123.a' :: double;
+
+-- '-2147483648' is the smallest int value
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/cast_op.sql:33]: I don't think we need so many tests tho.

Comment (HyukjinKwon) [reply sql/core/src/test/resources/sql-tests/inputs/cast_op.sql:33]: Can we just have few tests? A lot of tests are duplicates with `CastingSyntaxSuite` too

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/cast_op.sql:33]: maybe just add a few end-to-end tests in `cast.sql`? The underlying expression is the same so we probably don't need to test so many different values.

Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/inputs/cast_op.sql:33]: I agree, it is better to test more corner cases:
1. Chain of casts: `<expr> :: type1 :: type2`
2. Regular cast + new one: `CAST(<expr> :: type1 AS type2)`
3. Precedence of operators: `map[1] :: type1 [2]`
4. If you support only primary expr, add a negative test for non-primary expr
5. Incorrect taget type: `<expr> :: BINT`, `<expr> :: SELECT`

Comment (mitkedb) [reply sql/core/src/test/resources/sql-tests/inputs/cast_op.sql:33]: Addressed all the feedback
 - reduced the number of tests
 - consolidated tests in cast.sql
 - added tests suggested by Max

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -2128,6 +2128,15 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     }
   }
 
+  /**
+   * Create a [[Cast]] expression.
+   */
+  override def visitCastByColon(ctx: CastByColonContext): Expression = withOrigin(ctx) {
+    val rawDataType = typedVisit[DataType](ctx.dataType())
+    val dataType = CharVarcharUtils.replaceCharVarcharWithStringForCast(rawDataType)
+    Cast(expression(ctx.primaryExpression), dataType)
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2137]: +1

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/CastingSyntaxSuite.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  test("boolean expressions") {
+    assertEqual("(a and b) :: int", Cast('a && 'b, IntegerType))
+    assertEqual("(a or b) :: int", Cast('a || 'b, IntegerType))
+  }
+
+  test("arithmetic expressions") {
+    assertEqual("(a - b) :: int", Cast('a - 'b, IntegerType))
+    assertEqual("(a * b) :: int", Cast('a * 'b, IntegerType))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/CastingSyntaxSuite.scala:71]: can we also test `a + b::int` and make sure it produces `Add(a, Cast(b, int))`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 5]
```diff
@@ -2128,6 +2128,15 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     }
   }
 
+  /**
+   * Create a [[Cast]] expression.
```
[the remark this answers is not in the document]
Comment (mitkedb) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mitkedb) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2137]: fixed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/CastingSyntaxSuite.scala (hunk 4 above)
Comment (mitkedb) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/CastingSyntaxSuite.scala:71]: added.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Review by cloud-fan — APPROVED** [review]

**Review by beliefer — APPROVED** [review]

**Comment by MaxGekk** [comment]
This test failure seems related to your changes:
```
[info] - using _FUNC_ instead of function names in examples *** FAILED *** (7 milliseconds)
[info]   Expression class 'org.apache.spark.sql.catalyst.expressions.Cast' collection.this.IterableOnce.iterableOnceExtensionMethods[String](exampleRe.findAllIn(exprExamples)).toIterable.filter(((x$1: String) => setStmtRe.findFirstIn(x$1).isEmpty)).forall(((x$2: String) => x$2.contains("_FUNC_"))) was false (ExpressionInfoSuite.scala:125)
[info]   org.scalatest.exceptions.TestFailedException:
```
@mitkedb Please, take a look at it and fix.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 6]
```diff
@@ -438,11 +438,14 @@ object Cast extends QueryErrorsBase {
...
  */
 @ExpressionDescription(
-  usage = "_FUNC_(expr AS type) - Casts the value `expr` to the target data type `type`.",
+  usage = "_FUNC_(expr AS type) - Casts the value `expr` to the target data type `type`." +
+          " `expr` :: `type` alternative casting syntax is also supported.",
   examples = """
     Examples:
       > SELECT _FUNC_('10' as int);
        10
+      > SELECT '10' :: int;
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:447]: We usually use `_FUNC_`, BTW, but this case is special.

**Comment by mitkedb** [comment]
Debugging the remaining two unittest failures...

**Comment by mitkedb** [comment]
I can repro the failures with the diff, and they do not repro without the diff.

The following two unittest asserts are failing in QueryParsingErrorsSuite:

    checkError(
      exception = parseException("SELECT CAST(map('1',2) AS MAP<STRING>)"),
      errorClass = "PARSE_SYNTAX_ERROR",
      sqlState = "42601",
      parameters = Map("error" -> "'>'", "hint" -> ""))
      
      
      and 
      
          checkError(
      exception = parseException("SELECT CAST(struct(1,2,3) AS STRUCT<INT>)"),
      errorClass = "PARSE_SYNTAX_ERROR",
      sqlState = "42601",
      parameters = Map("error" -> "'>'", "hint" -> ""))
      
      
My hypothesis is that the SparkParserErrorStrategy result changed with the updates we made to the parser, and triggered a slightly different recovery error message. If this is the case, the fix is to just update the expected output to match what we're getting now. Unfortunately, this check is fairly fragile, but it is unclear to me how to fix it (and even if we want to fix, we should do it separately)

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryParsingErrorsSuite.scala [hunk 7]
```diff
@@ -630,7 +630,7 @@ class QueryParsingErrorsSuite extends QueryTest with SharedSparkSession with SQL
       exception = parseException("SELECT CAST(struct(1,2,3) AS STRUCT<INT>)"),
       errorClass = "PARSE_SYNTAX_ERROR",
       sqlState = "42601",
-      parameters = Map("error" -> "'>'", "hint" -> ""))
+      parameters = Map("error" -> "'<'", "hint" -> ": missing ')'"))
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryParsingErrorsSuite.scala:633]: Could you debug the `DefaultErrorStrategy` to see what is the root cause?

**Comment by MaxGekk** [comment]
> If this is the case, the fix is to just update the expected output to match what we're getting now.

I agree, let's fix the expected error params.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @mitkedb and @beliefer @cloud-fan @peter-toth @HyukjinKwon for review.

**Comment by MaxGekk** [comment]
@mitkedb Congratulations with your first contribution to Apache Spark!

**Pull request closed without merging**